qemu/tests/win_virtio_driver_update_by_installer.py

```diff
@@ -57,52 +52,78 @@ def change_virtio_media(cdrom_virtio):
-                                                         device_name_list,
-                                                         device_hwid_list):
+        for driver_name, device_name, device_hwid in zip(
+                    win_driver_installer_test.driver_name_list,
+                    win_driver_installer_test.device_name_list,
+                    win_driver_installer_test.device_hwid_list):
             win_driver_utils.install_driver_by_virtio_media(session, test,
                                                             devcon_path,
                                                             media_type,
                                                             driver_name,
                                                             device_hwid)
-        install_gagent(session, test, qemu_ga_pkg, gagent_install_cmd,
-                       gagent_pkg_info_cmd)
+        win_driver_installer_test.install_gagent(session, test, qemu_ga_pkg,
+                                                 gagent_install_cmd,
+                                                 gagent_pkg_info_cmd)
 
     error_context.context("Upgrade virtio driver to original",
                           test.log.info)
     change_virtio_media(params["cdrom_virtio"])
-    install_test_with_screen_on_desktop(vm, session, test,
+    win_driver_installer_test.install_test_with_screen_on_desktop(
+                                        vm, session, test,
                                         run_install_cmd,
                                         installer_pkg_check_cmd,
                                         copy_files_params=params)
     if params.get("need_reboot", "no") == "yes":
         session = vm.reboot(session)
 
-    check_gagent_version(session, test, gagent_pkg_info_cmd,
-                         expected_gagent_version)
-    driver_check(session, test, params)
+    error_context.context("Run driver function test after update",
+                          test.log.info)
+    driver_test_names = params["driver_test_names"].split()
+    del driver_test_names[1]
+    fail_tests = []
+    for test_name in driver_test_names:
+        test_func = "win_driver_installer_test.%s_test" % test_name
+        driver_test_params = params.get('driver_test_params_%s'
+                                        % test_name, '{}')
+        driver_test_params = ast.literal_eval(driver_test_params)
+        try:
+            eval("%s(test, params, vm, **driver_test_params)" % test_func)
+        except Exception:
+            fail_tests.append(test_name)
+    if fail_tests:
+        test.fail("Function test failed list is %s after update"
+                  % fail_tests)
+
+    win_driver_installer_test.check_gagent_version(session, test,
+                                                   gagent_pkg_info_cmd,
+                                                   expected_gagent_version)
+    win_driver_installer_test.driver_check(session, test, params)
```

Comment:
How about moving qga version check and driver check before function test.
update driver -> qga version check -> driver version check -> function test